### PR TITLE
Refactor `Compile.ps1` Script for Modularity and Enhanced Functionality

### PR DESCRIPTION
**Pull request by fam007e**
## Type of Change
- [x] Refactoring
- [x] New feature
- [x] Documentation update

## Description
This pull request refactors the `Compile.ps1` script with the following key improvements:

- **Moved `Update-Progress` Function:**  
  - The `Update-Progress` function has been moved to an external script for better modularity and reusability.
  
- **Added `Encode-JsonSpecialChars` Function:**  
  - A new function, `Encode-JsonSpecialChars`, has been introduced to handle JSON special character encoding more effectively. This improves the robustness of JSON handling and prevents potential formatting issues.
  
- **Implemented Error Handling:**  
  - Error handling has been enhanced with `try/catch` blocks during preprocessing and compilation steps, ensuring that issues are caught and handled gracefully.
  
- **Introduced Logging Functionality:**  
  - Logging has been added to track progress and capture errors during the script execution. This provides better traceability and helps in debugging.
  
- **Improved Path Handling:**  
  - Paths are now handled using `$PSScriptRoot` and `Join-Path` for better portability across different environments. This change improves the script's reliability when run in various directory structures.

## Testing
- The script was tested in both Windows PowerShell and PowerShell Core environments.
- The `-Debug`, `-Run`, and `-SkipPreprocessing` switches were tested to confirm that they function as expected.
- The syntax validation feature was tested to ensure that it catches common errors in the generated script.
- Logging was verified to ensure that all steps and errors are properly captured.

## Impact
- The refactoring enhances the maintainability and readability of the `Compile.ps1` script.
- The improved modularity allows for easier updates and potential reusability of components in other scripts.
- The introduction of error handling and logging makes the script more robust and easier to debug.

## Issue related to PR
- Resolves #

## Additional Information
- This refactoring is part of a larger effort to improve the overall code quality of the project. Future updates may include further modularization or additional features to enhance functionality.

## Checklist
- [x] My code adheres to the coding and style guidelines of the project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no errors/warnings/merge conflicts.

### CODE REVIEW

**Comment by Real-MullaC** [comment]
This only changes 1 file and does not add any files.

File: Compile.ps1 [hunk 1]
```diff
@@ -3,29 +3,39 @@ param (
...
 $sync = [Hashtable]::Synchronized(@{})
 $sync.PSScriptRoot = $workingdir
 $sync.configs = @{}
 
-function Update-Progress {
-    param (
-        [Parameter(Mandatory, position=0)]
-        [string]$StatusMessage,
+# Dot-source external functions
+. "$PSScriptRoot\tools\Update-Progress.ps1"
```
Comment (og-mrk) [inline Compile.ps1:18]: @fam007e you've stated in your PR description and in this line that there's indeed a script file with the name of `Update-Progress.ps1`, but in your **PR changes** there's no such file (you're only editing `Compile.ps1` Script).

Make sure to add this newly created file with Git CLI (or using GitHub Desktop if you're using it). You can add `tools/Update-Progress.ps1` file with the following git commands (if you've Git CLI installed):
```ps1
git add ./tools/Update-Progress.ps1
git commit -m "Add 'Update-Progress' Script Utility"
```

File: tools/Update-Progress.ps1 [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        [string]$Activity,
+
+        [Parameter(Position = 3)]
+        [switch]$LogProgress
+    )
+
+    # Default activity to "Processing" if not provided
+    if (-not $Activity) {
+        $Activity = "Processing"
+    }
```
Comment (og-mrk) [inline tools/Update-Progress.ps1:23]:
```suggestion
        [Parameter(Position = 2)]
        [string]$Activity = "Processing",  # Default activity to "Processing" if not provided

        [Parameter(Position = 3)]
        [switch]$LogProgress
    )
```
You can assign a parameter a default value, which'll be used if no value was passed when function is called. It's a standard feature in PowerShell, and IMO it makes the code shorter and more readable. plus I've left the comment to make it super clear that the default value will be `Processing`.

File: tools/Update-Progress.ps1 [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+    # Optionally log the progress to a file
+    if ($LogProgress) {
+        $logMessage = "{0:yyyy-MM-dd HH:mm:ss} - {1} - {2}% - {3}" -f (Get-Date), $Activity, $Percent, $StatusMessage
+        $logMessage | Out-File -FilePath "$PSScriptRoot\progress.log" -Append -Encoding utf8
+    }
+}
+
+# Example Usage:
+# Update-Progress "Processing files..." 50 "File Processing" -LogProgress
```
Comment (og-mrk) [inline tools/Update-Progress.ps1:36]: I recommend changing this documentation to be set in the beginning of `Update-Progress` function, it's considered PowerShell standards to do so, and follows the general coding style of this project. A quick example on this is the following:
```ps1
function Update-Progress {
    <#
        .SYNOPSIS
            A wrapper for PowerShell 'Write-Progress' Cmdlet.
        .PARAMETER StatusMessage
            A mandatory parameter which'll be used when displaying progress bar using 'Write-Progress' Cmdlet.
        .PARAMETER Percent
            # Fill in rest of documentation for parameters here...
        .EXAMPLE
            Update-Progress "Processing files..." 50 "File Processing" -LogProgress
    #>

    # Parameters here
    # ...

    # Rest of the code here
    # ...
}
```

Comment (fam007e) [reply tools/Update-Progress.ps1]: I have edited according to your criteria that I did not do before. Thanks for pointing it out.

**Review by og-mrk — CHANGES_REQUESTED** [review]
@fam007e Because of the new changes introduced in this PR, new `.log` files will be created when running `Compile.ps1` script.. which might accidently be added in the future, and because these logs are for error checking/diagnosis.. It's a good to ignore (not track) these files by git (_the version management system used by this project, and every project on GitHub_).

To do so (ignore `.log` files) you can add the following line in `.gitignore` file:
```
*.log
```

**Comment by og-mrk** [comment]
@fam007e I've fixed in [PR 1](https://github.com/fam007e/winutil/pull/1) (on your fork) the issues described in [this review comment](https://github.com/ChrisTitusTech/winutil/pull/2587#pullrequestreview-2257553116). Please consider reviewing my PR over on your fork, and test my changes after merging it. Thanks in advance 😄

**Comment by og-mrk** [comment]
@fam007e I would like to say, first and foremost.. thanks for your contribution, it's awesome to see others contribute to WinUtil, no matter how simple/complex their changes might be...

With that being said, and after some changes done to PR #2465 by @MyDrift-user (a somewhat medium sized Refactoring PR).. so the changes in this PR are not needed. I hope this doesn't discourages you from contributing to this project or other open source projects.. as WinUtil recently is in active development from several contributors, so conflict in interests may arise.

I recommend waiting to see @ChrisTitusTech opinion on this, but from a technical stand point these changes of yours will cause some problems, not to mention you'll need to update `Compile.ps1` script, that's.. if PR #2465 is merged in the first place.

**Pull request closed without merging**